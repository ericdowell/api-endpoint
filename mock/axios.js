export function BasicMock(options) {
  return new Promise((resolve, reject) => {
    if (options.errors) {
      return reject({
        config: {},
        response: {
          data: {
            errors: options.errors
          },
          status: options.status || 422
        }
      })
    }
    return resolve({
      config: {},
      data: { foo: 'bar' }
    })
  })
}
