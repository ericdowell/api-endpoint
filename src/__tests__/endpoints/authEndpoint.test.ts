import { AuthEndpoint } from '../..'
import axios from 'axios'
import { BasicMock } from '../mock/axios'

jest.spyOn(axios, 'request').mockImplementation(BasicMock)

const callEndpoint = (instance: AuthEndpoint): { [key: string]: Function } => {
  return {
    login: instance.login.bind(instance),
    logout: instance.logout.bind(instance),
    resetPassword: instance.resetPassword.bind(instance),
  }
}

describe(`${AuthEndpoint.name}`, (): void => {
  it('login method defaults to remember equal to true', async(): Promise<
    void
  > => {
    expect.assertions(2)
    const endpoint = new AuthEndpoint()
    const post = jest.spyOn(endpoint, 'post').mockImplementation((): any => true)
    await endpoint.login('email', 'password')
    expect(post).toHaveBeenCalledTimes(1)
    expect(post.mock.calls[0]).toMatchInlineSnapshot(`
      Array [
        "/login",
        Object {
          "data": Object {
            "email": "email",
            "password": "password",
            "remember": true,
          },
        },
      ]
    `)
  })

  it.each([
    ['login', 'post', ['email', 'password', false]],
    ['logout', 'post', []],
    ['resetPassword', 'post', ['email', 'token', 'password', 'password']],
  ])(
    'the %s calls %s parent method',
    async(method: string, calls, params): Promise<void> => {
      expect.assertions(2)
      const endpoint = new AuthEndpoint()
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      const methodCalled = jest.spyOn(endpoint, calls).mockImplementation((): any => true)
      await callEndpoint(endpoint)[method](...params)
      expect(methodCalled).toHaveBeenCalledTimes(1)
      // eslint-disable-next-line jest/prefer-inline-snapshots
      expect(methodCalled.mock.calls[0]).toMatchSnapshot()
    },
  )
})
