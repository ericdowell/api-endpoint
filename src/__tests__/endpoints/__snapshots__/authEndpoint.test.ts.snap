// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthEndpoint the login calls post parent method 1`] = `
Array [
  "/login",
  Object {
    "data": Object {
      "email": "email",
      "password": "password",
      "remember": false,
    },
  },
]
`;

exports[`AuthEndpoint the logout calls post parent method 1`] = `
Array [
  "/logout",
]
`;

exports[`AuthEndpoint the register calls post parent method 1`] = `
Array [
  "/register",
  Object {
    "data": Object {
      "email": "email",
      "email_confirmation": "email",
      "name": "John",
      "password": "password",
      "password_confirmation": "password",
      "remember": false,
    },
  },
]
`;

exports[`AuthEndpoint the requestPasswordReset calls post parent method 1`] = `
Array [
  "/password/email",
  Object {
    "data": Object {
      "email": "email",
    },
  },
]
`;

exports[`AuthEndpoint the resetPassword calls post parent method 1`] = `
Array [
  "/password/reset",
  Object {
    "data": Object {
      "email": "email",
      "password": "password",
      "password_confirmation": "password",
      "token": "token",
    },
  },
]
`;
